File: Python/Leetcode/Medium/Binary Search Tree to Greater Sum Tree.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:

    def bstToGst(self, root: TreeNode) -> TreeNode:
        global stackValue
        stackValue = 0 # 순회하면서 쌓는 점수

        #  right → root → left 순으로 순회
        def treeOrder(node: TreeNode):
            global stackValue
            if (node.right):
                treeOrder(node.right)

            stackValue += node.val
            node.val = stackValue

            if (node.left):
                treeOrder(node.left)

        treeOrder(root)
        return root

File: Java/Leetcode/src/Medium/BinarySearchTreeToGreaterSumTree.java
package Medium;

public class BinarySearchTreeToGreaterSumTree {

    int stackValue; // 순회하면서 쌓는 점수
    public TreeNode bstToGst(TreeNode root) {

        this.stackValue = 0; // 초기화
        treeOrder(root);

        return root;
    }

    // right → root → left 순으로 순회
    public void treeOrder(TreeNode node){

        if(node.right != null){
            treeOrder(node.right);
        }

        stackValue += node.val;
        node.val = stackValue;

        if(node.left != null){
            treeOrder(node.left);
        }

    }

    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode() {}
        TreeNode(int val) { this.val = val; }
        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }
}
